File: 프로그래머스/lv3/81303. 표 편집/표 편집.py
def solution(n, k, cmd):
    table = ["O"] * n
    remove = []

    graph = [[-1, 1]]
    for i in range(1, n-1):
        graph.append([i-1, i+1])
    graph.append([n-2, -1])

    for c in cmd:
        
        if c == "C":
            remove.append(k)
            table[k] = "X"
            node = graph[k]
            
            if node[1] == -1:
                k = node[0]
                graph[node[0]][1] = -1
            elif node[0] == -1:
                k = node[1]
                graph[node[1]][0] = -1
            else:
                graph[node[0]][1] = node[1]
                graph[node[1]][0] = node[0]
                k = node[1]
                
        elif c == "Z":
            idx = remove.pop()
            node = graph[idx]
            table[idx] = "O"
            graph[node[0]][1] = idx
            
            if node[1] != -1: graph[node[1]][0] = idx
            
            if k == -1: k = idx
            
        else:
            direc, num = c.split()
            num = int(num)
            
            if direc == "U":
                while num > 0:
                    k = graph[k][0]
                    num -= 1
                    
            else:
                while num > 0:
                    k = graph[k][1]
                    num -= 1

    return "".join(table)

File: 프로그래머스/lv3/81303. 표 편집/표 편집.java
import java.util.*;

class Solution {
    public String solution(int n, int k, String[] cmd) {
        String[] table = new String[n];
        Arrays.fill(table, "O");
        
        Stack<Node> remove = new Stack();
        
        Node current = set(n);
        while (k-- > 0)
            current = current.next;

        for (String c: cmd) {
            if (c.equals("C")) {
                remove.push(current);
                table[current.n] = "X";
                
                if (current.prev == null) {
                    current = current.next;
                    current.prev = null;
                } else if (current.next == null) {
                    current = current.prev;
                    current.next = null;
                } else {
                    current.prev.next = current.next;
                    current.next.prev = current.prev;
                    current = current.next;
                }
            } else if (c.equals("Z")) {
                Node node = remove.pop();
                table[node.n] = "O";
                
                if (node.prev != null)
                    node.prev.next = node;
                
                if (node.next != null)
                    node.next.prev = node;
                
                if (current == null)
                    current = node;

            } else {
                String[] tmp = c.split(" ");
                int num = Integer.parseInt(tmp[1]);
                
                if (tmp[0].equals("U")) {
                    while (num-- > 0)
                        current = current.prev;
                } else {
                    while (num-- > 0)
                        current = current.next;
                }
            }
        }
        return String.join("", table);
    }

    private Node set(int n) {
        Node head = new Node(0);
        Node prev = head;
        Node node;
        
        for (int i = 1; i < n; i++) {
            node = new Node(i);
            
            node.prev = prev;
            prev.next = node;
            
            prev = node;
        }
        return head;
    }

    class Node {
        int n;
        Node prev, next;
        
        public Node(int n) {
            this.n = n;
        }
    }
}